### PR TITLE
Implemented dshiftl(2)

### PR DESCRIPTION
**Pull request by parth121101**
Fixes: #3572

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -1066,6 +1067,57 @@ namespace Shiftl {
...
+        declare_basic_variables("_lcompilers_dshiftl_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("i", arg_types[0]);
+        fill_func_arg("j", arg_types[1]);
+        fill_func_arg("shift", arg_types[2]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        /*
+        * r = Dshiftl(x, y, shift)
+        * r = x << shift | y >> (32 - shift)
+        */
+        ASR::expr_t *cast = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[2], ASR::cast_kindType::IntegerToInteger, return_type, nullptr));
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_functions.h:1108]: This adds every time, how about we add this only for kind == 8?

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -1066,6 +1067,57 @@ namespace Shiftl {
...
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_dshiftl_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("i", arg_types[0]);
+        fill_func_arg("j", arg_types[1]);
+        fill_func_arg("shift", arg_types[2]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+        /*
+        * r = Dshiftl(x, y, shift)
+        * r = x << shift | y >> (32 - shift)
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_functions.h:1106]: Also, let's have the same implementation in both compile time and run time
```suggestion
        * r = x << shift | y >> (32 - shift) ! kind = 4
        or
        * r = x << shift | y >> (64 - shift) ! kind = 8
```

File: integration_tests/intrinsics_175.f90 [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    shift1 = 3
+    a = 7
+    b = 12
+    shift2 = 5
+
+    print *, dshiftl(10,4,3)
+    if(dshiftl(10,4,3) /= 80) error stop
+
+    print *, dshiftl(7_8,12_8,5)
+    if(dshiftl(7_8,12_8,5) /= 224) error stop
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/intrinsics_175.f90:15]:
```suggestion
    if(dshiftl(7_dp, 12_dp, 5) /= 224) error stop
```

File: integration_tests/intrinsics_175.f90 [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    integer(8) :: a, b, shift2
+    i = 10
+    j = 4
+    shift1 = 3
+    a = 7
+    b = 12
+    shift2 = 5
+
+    print *, dshiftl(10,4,3)
+    if(dshiftl(10,4,3) /= 80) error stop
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/intrinsics_175.f90:12]:
```suggestion
    if(dshiftl(10, 4, 3) /= 80) error stop
```

File: src/libasr/pass/intrinsic_functions.h [hunk 5]
```diff
@@ -1066,6 +1067,57 @@ namespace Shiftl {
...
+                ASR::ttype_t* t1, Vec<ASR::expr_t*> &args, diag::Diagnostics& diag) {
+        int64_t val1 = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
+        int64_t val2 = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+        int64_t shift = ASR::down_cast<ASR::IntegerConstant_t>(args[2])->m_n;
+        int kind1 = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_type);
+        int kind2 = ASRUtils::extract_kind_from_ttype_t(ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_type);
+        if(kind1 != kind2) {
+            append_error(diag, "The kind of first argument of 'dshiftl' intrinsic must be the same as second arguement", loc);
+            return nullptr;
+        }
```
[the remark this answers is not in the document]
Comment (parth121101) [reply src/libasr/pass/intrinsic_functions.h:1082]: Yes @Thirumalai-Shaktivel , for now I did this in the eval_dshiftl function itself.

**Review by certik — APPROVED** [review]
Fine with me, as long as @Thirumalai-Shaktivel is ok.

**Comment by parth121101** [comment]
@certik , @Thirumalai-Shaktivel I have resolved the conflicts. if you find this PR okay please merge this.

**Pull request merged**